src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,128 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private static final Object[] DEFAULTCAPACITY_EMPTY_ELEMENTDATA = {};
+    private T[] elementData;
+    private int size;
+
+    public ArrayList() {
+        this.elementData = (T[]) DEFAULTCAPACITY_EMPTY_ELEMENTDATA;
+    }
+
     @Override
     public void add(T value) {
-
+        resizeIfNeeded();
+        elementData[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("The specified index is"
+                    + " larger than the array size");
+        }
+        resizeIfNeeded();
 
+        for (int i = size; i > index; i--) {
+            elementData[i] = elementData[i - 1];
+        }
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        resizeIfNeeded();
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("The specified index is"
+                    + " larger than the array size");
+        }
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("The specified index is"
+                    + " larger than the array size");
+        }
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index >= size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("The specified index is"
+                    + " larger than the array size");
+        }
+        resizeIfNeeded();
+        T element = elementData[index];
+
+        for (int i = index; i < size; i++) {
+            elementData[i] = elementData[i + 1];
+        }
+        size--;
+        return element;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        boolean elementFound = false;
+
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] == null || elementData[i] == element) {
+                remove(i);
+                elementFound = true;
+                break;
+            } else if (elementData[i].equals(element)) {
+                remove(i);
+                elementFound = true;
+                break;
+            }
+        }
+
+        if (!elementFound) {
+            throw new NoSuchElementException("Element with such data was not found");
+        }
+        return element;
```

Comment:
1) there is null-safe comparison pattern, no need of if/else for it
2) we may use return from if statement to avoid additional if and variable
```suggestion
        for (int i = 0; i < size; i++) {
            if (element == elementData[i] || element != null && element.equals(elementData[i])) {
                return remove(i);
            }
        }
        throw new NoSuchElementException("Element with such data was not found");
```